### PR TITLE
Introduce the "comparable" relation

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
This pull request introduces a new relationship to check whether or not two types are "comparable to" one another. Specifically, it addresses the issues found in #5300 with union types and certain operations.

The new relationship generally subsumes the uses for assignability where we would check whether or not two types were assignable to or assignable from one another. These uses were in checking for compatibility between

* the types of the left and right sides of equality/inequality operators (`===`, `!==`, `==`, `!=`).
* the types of `case` clause expressions and their respective `switch` expressions.
* the type of an expression in a type assertion with the type being asserted.

Like with other type relations, the comparable relation takes a *source* and a *target*. It is the most lenient relation, so types that are assignable are always comparable, but the reverse is not necessarily true. The comparable relation differs from the assignability relation in that if the *source* is a union type, the relation succeeds provided that each of its constituents are comparable to the *target*.

### CODE REVIEW

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -888,6 +888,10 @@
         "category": "Error",
         "code": 2322
     },
+    "Type '{0}' is not comparable with type '{1}'.": {
```
Comment (sandersn) [inline src/compiler/diagnosticMessages.json]: I prefer 'comparable to' here. A little research shows that grammarians think they're about equally common, while Google's hideously unreliable hit count shows 4 times more hits for 'comparable to'*.

But the function is also named `isTypeComparableTo`, so I think there's precedent. :)

*Google hit counts are the laziest way to do quantitative linguistics.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -12689,12 +12725,12 @@ namespace ts {
 
                 if (produceDiagnostics && clause.kind === SyntaxKind.CaseClause) {
                     let caseClause = <CaseClause>clause;
-                    // TypeScript 1.0 spec (April 2014):5.9
+                    // TypeScript 1.0 spec (April 2014): 5.9
                     // In a 'switch' statement, each 'case' expression must be of a type that is assignable to or from the type of the 'switch' expression.
```
Comment (sandersn) [inline src/compiler/checker.ts]: update 'assignable' to 'comparable'

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -4746,6 +4755,14 @@ namespace ts {
             return checkTypeRelatedTo(source, target, assignableRelation, errorNode, headMessage, containingMessageChain);
         }
 
+        /**
+         * This is *not* a bi-directional relationship.
+         * If one needs to check both directions for comparability, use a second call to this function or 'isTypeComparableTo'.
+         */
+        function checkTypeComparableTo(source: Type, target: Type, errorNode: Node, headMessage?: DiagnosticMessage, containingMessageChain?: DiagnosticMessageChain): boolean {
```
Comment (sandersn) [inline src/compiler/checker.ts]: nit: This function is hard to read on a portrait (1200 px) monitor with github's generous margins cutting it down to about 700 px. Might not be worth fixing.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -4738,6 +4739,14 @@ namespace ts {
             return checkTypeAssignableTo(source, target, /*errorNode*/ undefined);
         }
 
+        /**
+         * This is *not* a bi-directional relationship.
```
Comment (sandersn) [inline src/compiler/checker.ts]: Is this true of `isTypeAssignableTo` too? Its former usage in this diff indicates 'yes'.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -4738,6 +4739,14 @@ namespace ts {
             return checkTypeAssignableTo(source, target, /*errorNode*/ undefined);
         }
 
+        /**
+         * This is *not* a bi-directional relationship.
+         * If one needs to check both directions for comparability, use a second call to this function or 'checkTypeComparableTo'.
```
Comment (sandersn) [inline src/compiler/checker.ts]: All uses *do* need to check both directions, so probably you should have a version that does that. You might not even need the uni-directional ones except for consistency with the other relations.

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -4859,12 +4884,21 @@ namespace ts {
 
                 // Note that the "each" checks must precede the "some" checks to produce the correct results
                 if (source.flags & TypeFlags.Union) {
-                    if (result = eachTypeRelatedToType(<UnionType>source, target, reportErrors)) {
+                    if (relation === comparableRelation) {
+                        result = someTypeRelatedToType(source as UnionType, target, reportErrors);
```
Comment (sandersn) [inline src/compiler/checker.ts]: Please choose one cast syntax and make at least the changed lines use it, preferably something like all the casts in the function.

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -4859,12 +4884,21 @@ namespace ts {
 
                 // Note that the "each" checks must precede the "some" checks to produce the correct results
```
Comment (sandersn) [inline src/compiler/checker.ts]: This comment needs to reflect the more complex checks.

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -4873,7 +4907,8 @@ namespace ts {
                     // on either side that need to be prioritized. For example, A | B = (A | B) & (C | D) or
                     // A & B = (A & B) | (C & D).
                     if (source.flags & TypeFlags.Intersection) {
-                        // If target is a union type the following check will report errors so we suppress them here
+                        // If target is a union type then the check following this one will report errors,
```
Comment (sandersn) [inline src/compiler/checker.ts]: I don't think this improves on the old wording.

File: src/compiler/checker.ts (hunk 4 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: It is true of `isTypeAssignableTo` but `isTypeAssignableTo` has a clear direction; on the other hand, `isTypeComparableTo` is vague enough that it warrants a comment. One would think that `A` being comparable to `B` implies that `B` is comarable to `A`, but that's not quite the case.

I *could* have made `isTypeComparableTo` a bidirectional relation, but I decided not to based on the way in which we use the relation in type assertions. Specifically, with type assertions, we check whether the expression type is comparable to the asserted type and that the asserted type is comparable to the *widened* type of the expression.

Actually, I'm not sure if there is any reason we don't check against the widened type for both. I'll experiment with it.

File: src/compiler/checker.ts (hunk 5 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: See [this comment](https://github.com/Microsoft/TypeScript/pull/5517#discussion_r43935354)

File: src/compiler/checker.ts (hunk 8 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: It does if you have more context. "The following check" implies "the one directly below"

File: src/compiler/diagnosticMessages.json (hunk 1 above)
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: Yeah, I couldn't decide, but I'm okay with either.

File: src/compiler/checker.ts (hunk 4 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: As discussed offline, I think that's a bug. I will change that in a later PR

File: tests/baselines/reference/literals-negative.errors.txt [hunk 9]
```diff
@@ -8,7 +8,7 @@ tests/cases/compiler/literals-negative.ts(5,9): error TS2352: Neither type 'numb
     var s = <string>(null);
     var b = <boolean>(n);
             ~~~~~~~~~~~~
-!!! error TS2352: Neither type 'number' nor type 'boolean' is assignable to the other.
+!!! error TS2323: Type 'number' is not comparable to type 'boolean'.
```
Comment (RyanCavanaugh) [inline tests/baselines/reference/literals-negative.errors.txt]: "comparable" is an odd error to issue when complaining about a type assertion. Can we say "x cannot be converted to y" with the "comparable to" error as the next diagnostic chain?

Comment (DanielRosenwasser) [inline tests/baselines/reference/literals-negative.errors.txt]: Yes.

**Comment by mhegazy** [comment]
@DanielRosenwasser can we close this.

**Comment by DanielRosenwasser** [comment]
I'm planning to merge this in for the tests as a heads up.

**Pull request merged**